File: SIMS/DOM/DataListTemplate.vb
Imports Microsoft.VisualBasic
Imports System.Data
Imports System.Net

Public Class DataListTemplate
    Implements ITemplate

    Dim _itemType As ListItemType
    Private _no As String
    Private _view As String
    Private _recordType As String
    Private _wy As String
    Private _count As Integer

    Public Property NoOfPeriods() As String
        Get
            Return _no
        End Get
        Set(ByVal Value As String)
            _no = Value
        End Set
    End Property

    Public Property WY() As String
        Get
            Return _wy
        End Get
        Set(ByVal value As String)
            _wy = value
        End Set
    End Property

    Public Property RecordType() As String
        Get
            Return _recordType
        End Get
        Set(ByVal Value As String)
            _recordType = Value
        End Set
    End Property

    Public Property Count() As Integer
        Get
            Return _count
        End Get
        Set(ByVal value As Integer)
            _count = value
        End Set
    End Property

    Public Property WhichView() As String
        Get
            Return _view
        End Get
        Set(ByVal value As String)
            _view = value
        End Set
    End Property

    Public Sub New(ByVal Type As ListItemType)
        _itemType = Type
    End Sub

    Public Sub InstantiateIn(ByVal container As System.Web.UI.Control) Implements ITemplate.InstantiateIn
        Dim lc As Literal = New Literal()

        Select Case _itemType
            Case ListItemType.Header

                lc.Text = "<fieldset style=""width:340px;""><legend>" & RecordType & "</legend><div style=""padding:5px;"">" & _
                    "<table cellpadding=""3"" style=""border: 1px solid #863d02;background-color:white;"">" & _
                    "  <tr>" & _
                    "    <td style=""background-color:#863d02;color:white;text-align:center;"">Period ID</td>" & _
                    "    <td style=""background-color:#863d02;color:white;text-align:center;"">Begin Date</td>" & _
                    "    <td style=""background-color:#863d02;color:white;text-align:center;"">End Date</td>" & _
                    "    <td style=""background-color:#863d02;color:white;text-align:center;"">Status</td>" & _
                    "  </tr>"

                Count = 1

                container.Controls.Add(lc)
            Case ListItemType.Item

                Count += 1

                AddHandler lc.DataBinding, AddressOf TemplateControl_DataBinding

                container.Controls.Add(lc)
            Case ListItemType.Footer

                If WhichView = "dates" Then
                    lc.Text = "</table>" & _
                        "<div style=""text-align:right;padding-top:10px;"">showing the most recent record period(s)</div></div></fieldset><br /><br />"
                Else
                    lc.Text = "</table>" & _
                        "<div style=""text-align:right;padding-top:10px;"">showing record period(s) for the " & WY & " WY</div></div></fieldset><br /><br />"
                End If

                container.Controls.Add(lc)
        End Select

    End Sub

    Private Sub TemplateControl_DataBinding(ByVal sender As Object, ByVal e As System.EventArgs)
        Dim lc As Literal

        lc = CType(sender, Literal)

        Dim container As DataListItem = CType(lc.NamingContainer, DataListItem)

        Dim status_va As String = DataBinder.Eval(container.DataItem, "status_va")
        Dim period_id As Integer = DataBinder.Eval(container.DataItem, "period_id")
        Dim lock_img As String
        Dim locked As Boolean

        Dim p As New RecordPeriod(period_id)

        locked = p.PeriodIsLocked

        If locked Then
            lock_img = " &nbsp;<img src=""images/lock.png"" alt=""LOCKED"" />"
        Else
            lock_img = ""
        End If

        If WhichView = "dates" Then
            If Not locked Then
                If Count = CInt(NoOfPeriods) Then
                    lc.Text += "<tr><td>" & period_id & "</td><td><a href=""EditPeriodDates.aspx?period_id=" & _
                        period_id & "&dt=beg2"">" & DataBinder.Eval(container.DataItem, "period_beg_dt") & "</a>" & _
                        "</td><td><a href=""EditPeriodDates.aspx?period_id=" & period_id & "&dt=end2"">" & _
                        DataBinder.Eval(container.DataItem, "period_end_dt") & "</a></td><td>" & _
                        status_va & "</td></tr>"
                Else
                    lc.Text += "<tr><td>" & period_id & "</td><td>" & _
                        DataBinder.Eval(container.DataItem, "period_beg_dt") & _
                        "</td><td><a href=""EditPeriodDates.aspx?period_id=" & period_id & "&dt=end1"">" & _
                        DataBinder.Eval(container.DataItem, "period_end_dt") & "</a></td><td>" & _
                        status_va & "</td></tr>"
                End If
            Else
                lc.Text += "<tr><td style=""color:#CCCCCC;"">" & period_id & "</td><td style=""color:#CCCCCC;"">" & _
                    DataBinder.Eval(container.DataItem, "period_beg_dt") & "</td><td style=""color:#CCCCCC;"">" & DataBinder.Eval(container.DataItem, "period_end_dt") & _
                    "</td><td style=""color:#CCCCCC;"">" & status_va & lock_img & "</td></tr>"
            End If
        Else
            If Count = CInt(NoOfPeriods) Then
                If Not status_va = "Working" And Not status_va = "Checking" And Not status_va = "Reviewing" And Not status_va = "Rework" And Not status_va = "Reworking" And Not locked Then
                    lc.Text += "<tr><td>" & period_id & "</td><td>" & DataBinder.Eval(container.DataItem, "period_beg_dt") & _
                        "</td><td>" & DataBinder.Eval(container.DataItem, "period_end_dt") & "</td><td><a href=""EditPeriodStatus.aspx?period_id=" & DataBinder.Eval(container.DataItem, "period_id") & _
                        """>" & status_va & "</a></td></tr>"
                Else
                    lc.Text += "<tr><td style=""color:#CCCCCC;"">" & period_id & "</td><td style=""color:#CCCCCC;"">" & _
                        DataBinder.Eval(container.DataItem, "period_beg_dt") & "</td><td style=""color:#CCCCCC;"">" & DataBinder.Eval(container.DataItem, "period_end_dt") & _
                        "</td><td style=""color:#CCCCCC;"">" & status_va & lock_img & "</td></tr>"
                End If
            Else
                lc.Text += "<tr><td style=""color:#CCCCCC;"">" & period_id & "</td><td style=""color:#CCCCCC;"">" & _
                    DataBinder.Eval(container.DataItem, "period_beg_dt") & "</td><td style=""color:#CCCCCC;"">" & DataBinder.Eval(container.DataItem, "period_end_dt") & _
                    "</td><td style=""color:#CCCCCC;"">" & status_va & lock_img & "</td></tr>"
            End If
        End If




    End Sub

End Class

File: Core/DataListTemplate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Web.UI;
using System.Web.UI.WebControls;

namespace Core
{
    public class DataListTemplate : ITemplate
    {
        private Data.SIMSDataContext db = new Data.SIMSDataContext();

        ListItemType _itemType;
        private string _no;
        private string _view;
        private string _recordType;
        private string _wy;

        private int _count;
        public string NoOfPeriods
        {
            get { return _no; }
            set { _no = value; }
        }

        public string WY
        {
            get { return _wy; }
            set { _wy = value; }
        }

        public string RecordType
        {
            get { return _recordType; }
            set { _recordType = value; }
        }

        public int Count
        {
            get { return _count; }
            set { _count = value; }
        }

        public string WhichView
        {
            get { return _view; }
            set { _view = value; }
        }

        public DataListTemplate(ListItemType Type)
        {
            _itemType = Type;
        }

        public void InstantiateIn(System.Web.UI.Control container)
        {
            Literal lc = new Literal();

            switch (_itemType)
            {
                case ListItemType.Header:

                    lc.Text = "<fieldset style=\"width:340px;\"><legend>" + RecordType + "</legend><div style=\"padding:5px;\">" + 
                        "<table cellpadding=\"3\" style=\"border: 1px solid #863d02;background-color:white;\">" + 
                        "  <tr>" + 
                        "    <td style=\"background-color:#863d02;color:white;text-align:center;\">Period ID</td>" + 
                        "    <td style=\"background-color:#863d02;color:white;text-align:center;\">Begin Date</td>" + 
                        "    <td style=\"background-color:#863d02;color:white;text-align:center;\">End Date</td>" + 
                        "    <td style=\"background-color:#863d02;color:white;text-align:center;\">Status</td>" + 
                        "  </tr>";

                    Count = 1;

                    container.Controls.Add(lc);
                    break;
                case ListItemType.Item:

                    Count += 1;

                    lc.DataBinding += TemplateControl_DataBinding;

                    container.Controls.Add(lc);
                    break;
                case ListItemType.Footer:

                    if (WhichView == "dates")
                    {
                        lc.Text = "</table>" + 
                            "<div style=\"text-align:right;padding-top:10px;\">showing the most recent record period(s)</div></div></fieldset><br /><br />";
                    }
                    else
                    {
                        lc.Text = "</table>" + 
                            "<div style=\"text-align:right;padding-top:10px;\">showing record period(s) for the " + WY + " WY</div></div></fieldset><br /><br />";
                    }

                    container.Controls.Add(lc);
                    break;
            }

        }

        private void TemplateControl_DataBinding(object sender, System.EventArgs e)
        {
            Literal lc = default(Literal);

            lc = (Literal)sender;

            DataListItem container = (DataListItem)lc.NamingContainer;

            string status_va = DataBinder.Eval(container.DataItem, "status_va").ToString();
            int period_id = Convert.ToInt32(DataBinder.Eval(container.DataItem, "period_id"));
            string lock_img = null;
            bool locked = false;

            var period = db.RecordAnalysisPeriods.FirstOrDefault(p => p.period_id == period_id);

            if (period.Record.RecordLock != null) locked = true;

            if (locked)
            {
                lock_img = " &nbsp;<img src=\"../images/lock.png\" alt=\"LOCKED\" />";
            }
            else
            {
                lock_img = "";
            }

            if (WhichView == "dates")
            {
                if (!locked)
                {
                    if (Count == Convert.ToInt32(NoOfPeriods))
                    {
                        lc.Text += "<tr><td>" + period_id + "</td><td><a href=\"PeriodDates.aspx?period_id=" + period_id + "&dt=beg2\">" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_beg_dt")) + "</a>" + "</td><td><a href=\"PeriodDates.aspx?period_id=" + period_id + "&dt=end2\">" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_end_dt")) + "</a></td><td>" + status_va + "</td></tr>";
                    }
                    else
                    {
                        lc.Text += "<tr><td>" + period_id + "</td><td>" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_beg_dt")) + "</td><td><a href=\"PeriodDates.aspx?period_id=" + period_id + "&dt=end1\">" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_end_dt")) + "</a></td><td>" + status_va + "</td></tr>";
                    }
                }
                else
                {
                    lc.Text += "<tr><td style=\"color:#CCCCCC;\">" + period_id + "</td><td style=\"color:#CCCCCC;\">" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_beg_dt")) + "</td><td style=\"color:#CCCCCC;\">" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_end_dt")) + "</td><td style=\"color:#CCCCCC;\">" + status_va + lock_img + "</td></tr>";
                }
            }
            else
            {
                if (Count == Convert.ToInt32(NoOfPeriods))
                {
                    if (!(status_va == "Analyzing") & !(status_va == "Approving") & !(status_va == "Reanalyze") & !locked)
                    {
                        lc.Text += "<tr><td>" + period_id + "</td><td>" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_beg_dt")) + "</td><td>" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_end_dt")) + "</td><td><a href=\"PeriodStatus.aspx?period_id=" + DataBinder.Eval(container.DataItem, "period_id") + "\">" + status_va + "</a></td></tr>";
                    }
                    else
                    {
                        lc.Text += "<tr><td style=\"color:#CCCCCC;\">" + period_id + "</td><td style=\"color:#CCCCCC;\">" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_beg_dt")) + "</td><td style=\"color:#CCCCCC;\">" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_end_dt")) + "</td><td style=\"color:#CCCCCC;\">" + status_va + lock_img + "</td></tr>";
                    }
                }
                else
                {
                    lc.Text += "<tr><td style=\"color:#CCCCCC;\">" + period_id + "</td><td style=\"color:#CCCCCC;\">" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_beg_dt")) + "</td><td style=\"color:#CCCCCC;\">" + String.Format("{0:MM/dd/yyyy}", DataBinder.Eval(container.DataItem, "period_end_dt")) + "</td><td style=\"color:#CCCCCC;\">" + status_va + lock_img + "</td></tr>";
                }
            }




        }

    }
}
